x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_initialize_rules_table_saved_state.test.ts

```diff
@@ -0,0 +1,427 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { renderHook } from '@testing-library/react-hooks';
+import { useKibana } from '../../../../../common/lib/kibana';
+import { useInitializeUrlParam } from '../../../../../common/utils/global_query_string';
+import { RULES_TABLE_MAX_PAGE_SIZE } from '../../../../../../common/constants';
+import { useInitializeRulesTableSavedState } from './use_initialize_rules_table_saved_state';
+import type { RulesTableActions } from './rules_table_context';
+import {
+  INITIAL_FILTER_OPTIONS,
+  INITIAL_SORTING_OPTIONS,
+  useRulesTableContext,
+} from './rules_table_context';
+import type { RulesTableSavedState } from './rules_table_saved_state';
+
+jest.mock('../../../../../common/lib/kibana');
+jest.mock('../../../../../common/utils/global_query_string');
+jest.mock('./rules_table_context', () => ({
+  INITIAL_FILTER_OPTIONS: jest.requireActual('./rules_table_context').INITIAL_FILTER_OPTIONS,
+  INITIAL_SORTING_OPTIONS: jest.requireActual('./rules_table_context').INITIAL_SORTING_OPTIONS,
+  useRulesTableContext: jest.fn(),
+}));
+
+function mockState({
+  urlState,
+  storageState,
+}: {
+  urlState: RulesTableSavedState | null;
+  storageState: RulesTableSavedState | null;
+}): void {
+  (useInitializeUrlParam as jest.Mock).mockImplementation(
+    (_, cb: (params: RulesTableSavedState | null) => void) => cb(urlState)
+  );
+  (useKibana as jest.Mock).mockReturnValue({
+    services: { sessionStorage: { get: jest.fn().mockReturnValue(storageState) } },
+  });
+}
+
+describe('useInitializeRulesTableSavedState', () => {
+  const urlSavedState: RulesTableSavedState = {
+    searchTerm: 'test',
+    showCustomRules: true,
+    tags: ['test'],
+    sort: {
+      field: 'name',
+      order: 'asc',
+    },
+    page: 2,
+    perPage: 10,
+  };
+  const storageSavedState: RulesTableSavedState = {
+    searchTerm: 'test',
+    showCustomRules: true,
+    tags: ['test'],
+    sort: {
+      field: 'name',
+      order: 'asc',
+    },
+    page: 1,
+    perPage: 20,
+  };
+  let actions: Partial<RulesTableActions>;
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+
+    actions = {
+      setFilterOptions: jest.fn(),
+      setSortingOptions: jest.fn(),
+      setPage: jest.fn(),
+      setPerPage: jest.fn(),
+    };
+
+    (useRulesTableContext as jest.Mock).mockReturnValue({ actions });
```

Comment:
Unfortunately I need to make expectations on actions and mock actions from `__mock__` folder just don't work. It's clear the mocks are invoked but the tests fail with the error the mocks aren't invoked. The problem lays under the hood, when mocks are exported from another module it seems a new instance is created which breaks tests.

@xcrzx If you know the solution or have another thoughts on this topic it would be great.

Reply:
Okay, then I think it would be better to mock the entire context with `useRulesTableContextMock.create()`. That will help us keep the mocked object structure in one place, and if something changes, we'll need to update only there. Plus, that mock is strongly typed, which makes it easier to find affected code when types change.

See how a test could be implemented in that case, for example:

https://github.com/elastic/kibana/blob/2efe77b5d847797c6d4763d540a117044fd69a81/x-pack/plugins/security_solution/public/detections/components/rules/rule_switch/index.test.tsx#L92-L106